google/cloud/sql/connector/connector.py

```diff
@@ -234,6 +234,11 @@ async def connect_async(
         # helper function to wrap in timeout
         async def get_connection() -> Any:
             instance_data, ip_address = await instance.connect_info(ip_type)
+
+            # validate `user` param for automatic IAM database authn
+            if enable_iam_auth:
+                validate_database_user(instance_data.database_version, kwargs["user"])
```

Comment:
Drive by comment from a user of this library - hope y'all don't mind.

It would be really nice if `enable_iam_auth` is set, we do only a little bit of extra work to make #310 happen and format the user for the caller.

Sample code example
```python
if enable_iam_auth:
  if "user" not in kwargs:
    if not hasattr(self._credentials, "service_account_email"):
      # Tell the caller to either explicitly pass in user, or pass in credentials auth to service account.
      raise SomeError("some message")
    user = format_user_for_sql(instance_data.database_version, self._credentials.service_account_email)
  else:
    user = kwargs["user"]
    validate_database_user(instance_data.database_version, user)
```

Reply:
I just figured if all the work is already being done to figure out what is appropriate for each dialect, it would be nice to just to it for the caller, so they don't have to duplicate this logic on their end.

Reply:
Hi @aebrahim, thanks for the comment! We always welcome ideas and suggestions for the connector 😄  

This PR in general is meant to be a step in the right direction towards what you have outlined here in the comment and was inspired by your original #310. I think it would be an awesome feature to have!

I've been looking into different ways to make the `user` optional. The main reason we aren't ready to pull the trigger on an implementation like the one you proposed just yet is because we want to maintain feature parity. Not only across different credential types, (i.e. service account vs. user credentials vs. application credentials) but also across our connector libraries. We try to maintain feature parity across this library, the [Go connector](https://github.com/GoogleCloudPlatform/cloud-sql-go-connector) and the [Java connector](https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory). While the feature is fairly straightforward to implement here in Python, the other languages aren't quite as easy.

We want to make the user experience as consistent as possible across the different languages. I will bring this back up with the team to discuss if diverging slightly here and offering this feature only in Python can be done, because I do see the immense value and improved user experience of the feature.

Thanks again for your enthusiasm and suggestion, they are always appreciated. I will hopefully have an update soon for you on this.

Reply:
Thanks for the detailed reasoning (as well as all the hard work on this package) @jackwotherspoon!

Reply:
@aebrahim based on your comment here we have updated the PR to actually format the `user` arg and not just validate it. This will abstract away the need to format MySQL and Postgres IAM users differently and allow developers to just pass in the full user or service account email as the `user` arg when using the Connector with IAM authentication.

This is hopefully a better step in the right direction to making the `user` arg optional down the road. Will try and prioritize this in the new year!

Thanks for the wonderful suggestions and feedback!